cpp/include/raft/matrix/gather.cuh

```diff
@@ -44,13 +47,13 @@ namespace raft::matrix {
  * @param  out          Pointer to the output matrix (assumed to be row-major)
  * @param  stream       CUDA stream to launch kernels within
  */
-template <typename MatrixIteratorT, typename MapIteratorT>
-void gather(const MatrixIteratorT in,
-            int D,
-            int N,
-            MapIteratorT map,
-            int map_length,
-            MatrixIteratorT out,
+template <typename InputIteratorT, typename MapIteratorT, typename OutputIteratorT, typename IndexT>
+void gather(const InputIteratorT in,
+            IndexT D,
+            IndexT N,
+            const MapIteratorT map,
+            IndexT map_length,
+            OutputIteratorT out,
             cudaStream_t stream)
```

Comment:
Also, does `gather` compare well against `copy_selected` in terms of speed?

Reply:
I found a lot of issues with `matrix::gather`, from compilation errors to poor performance, so I opened a separate PR: #1126

Reply:
> Do I understand it right, that since you split MatrixIteratorT into InputIteratorT and OutputIteratorT, we can use this in place of knn::details:utils::copy_selected? Then, could you please remove copy_selected and replace the use sites?

@achirkin I just noticed that in IVF-PQ it is used with `ld_src != ld_dst`. `matrix::gather` doesn't support that yet. I'll replace all the use sites I can, but we'll have to keep `copy_selected` for now if we don't want to block the PR.